src/main/java/org/jabref/gui/actions/ExportLinkedFilesAction.java

```diff
@@ -0,0 +1,164 @@
+import org.jabref.gui.util.DirectoryDialogConfiguration;
+import org.jabref.logic.l10n.Localization;
+import org.jabref.logic.util.OS;
+import org.jabref.logic.util.io.FileUtil;
+import org.jabref.model.database.BibDatabaseContext;
+import org.jabref.model.entry.BibEntry;
+import org.jabref.model.entry.LinkedFile;
+import org.jabref.model.util.FileHelper;
+import org.jabref.model.util.OptionalUtil;
+import org.jabref.preferences.JabRefPreferences;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.controlsfx.dialog.ProgressDialog;
+
+public class ExportLinkedFilesAction extends AbstractAction {
+
+    private static final Log LOGGER = LogFactory.getLog(ExportLinkedFilesAction.class);
+    private final BiFunction<Path, Path, Path> resolvePathFilename = (path, file) -> {
+        return path.resolve(file.getFileName());
+    };
+    private final DialogService ds = new FXDialogService();
+    private long totalFilesCount;
+    private BibDatabaseContext databaseContext;
+    private Optional<Path> exportPath = Optional.empty();
+    private List<BibEntry> entries;
+
+    public ExportLinkedFilesAction() {
+        super(Localization.lang("Copy attached files to folder..."));
+    }
+
+    @Override
+    public void actionPerformed(ActionEvent e) {
+
+        DirectoryDialogConfiguration dirDialogConfiguration = new DirectoryDialogConfiguration.Builder()
+                .withInitialDirectory(Paths.get(Globals.prefs.get(JabRefPreferences.EXPORT_WORKING_DIRECTORY)))
+                .build();
+        entries = JabRefGUI.getMainFrame().getCurrentBasePanel().getSelectedEntries();
+        exportPath = DefaultTaskExecutor
+                .runInJavaFXThread(() -> ds.showDirectorySelectionDialog(dirDialogConfiguration));
+
+        exportPath.ifPresent(path -> {
+            databaseContext = JabRefGUI.getMainFrame().getCurrentBasePanel().getDatabaseContext();
+            totalFilesCount = entries.stream().flatMap(entry -> entry.getFiles().stream()).count();
+
+            Service<Void> exportService = new ExportService();
+            startServiceAndshowProgessDialog(exportService);
+        });
+    }
+
+    private <V> void startServiceAndshowProgessDialog(Service<V> service) {
+        DefaultTaskExecutor.runInJavaFXThread(() -> {
+            service.start();
+
+            ProgressDialog progressDialog = new ProgressDialog(service);
+            progressDialog.setOnCloseRequest(evt -> service.cancel());
+            DialogPane dialogPane = progressDialog.getDialogPane();
+            dialogPane.getButtonTypes().add(ButtonType.CANCEL);
+            Button cancelButton = (Button) dialogPane.lookupButton(ButtonType.CANCEL);
+            cancelButton.setOnAction(evt -> {
+                service.cancel();
+                progressDialog.close();
+            });
+            progressDialog.showAndWait();
+        });
+    }
+
+    private class ExportService extends Service<Void> {
```

Comment:
This is rather nitpicky, but what about extracting this into a separate class file in the same package with default visibility? You could also extract parts of `call()` into separate private methods. All this wouldn't change anything about the functionality, but it would make the code a little less chunky and more readable.

Reply:
I see no reason why I should extract that in a new class. It's just a single Service for one purpose.
But I think I can surely extract some private methods

Reply:
One reason is that you could (and should) write tests for the `ExportService`.